### PR TITLE
Savestate Manager: Break dependency on DialogSelect and add "Saved with"

### PR DESCRIPTION
**Pull request by garbear**
## Description

This PR breaks the dependency of the savestate manager on DialogSelect, and adds a new info thumb - the emulator that created the save.

This is important, because currently there's no way in-game to see which emulator each save belongs to. It fixes a usability issue, because I can easily see this creating much confusion, such as "Where are my saves??" in-game when accidentally choosing a different emulator.

Mechanically, the process was to copy all the stuff we needed from DialogSelect into DialogGameSaves, and then improve the logic and test test test.

I also took the opportunity to totally re-do the dialog's XML, changing it to show 3x2 savestates instead of 2x2, which is more suitable for a 16:9 display.

## Motivation and context

"Saved with" seemed like such a simple addition. But when I went to add the feature, I immediately had to write a hack for DialogSelect and was like nahhhh......

It was beneficial to use DialogSelect initially because it offered everything we needed without extra code, but in hindsight we should have anticipated the problems with having half your dialog code spread elsewhere in the codebase. Sometimes moderate duplication is desirable.

This is my final expected contribution to savestates in v20.1.

## How has this been tested?

Tested in my latest round of builds: https://github.com/garbear/xbmc/releases

## What is the effect on users?

* Shows the savestate's emulator when choosing a savestate

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

File: xbmc/games/addons/GameClient.cpp [hunk 1]
```diff
@@ -694,3 +696,34 @@ bool CGameClient::cb_input_event(KODI_HANDLE kodiInstance, const game_input_even
 
   return gameClient->Input().ReceiveInputEvent(*event);
 }
+
+void CGameClient::ParseLibretroName(const std::string& addonName,
```
Comment (lrusak) [inline xbmc/games/addons/GameClient.cpp:700]: what about
```suggestion
std::pair<std::string, std::string> CGameClient::ParseLibretroName(const std::string& addonName,
```

Then you can use:
```
[m_emulatorName, m_platforms] = ParseLibretroName(Name());
```

**Review by lrusak — APPROVED** [review]
Nice improvement and abstraction.

File: xbmc/games/addons/GameClient.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/games/addons/GameClient.cpp:700]: Me likey

Comment (garbear) [reply xbmc/games/addons/GameClient.cpp:700]: @lrusak Can you help with your suggestion? I make the following changes:

```patch
--- a/xbmc/games/addons/GameClient.h
+++ b/xbmc/games/addons/GameClient.h
@@ -18,6 +18,7 @@
 #include <set>
 #include <stdint.h>
 #include <string>
+#include <utility>
 
 class CFileItem;
 
@@ -214,7 +215,23 @@ private:
   static bool cb_input_event(KODI_HANDLE kodiInstance, const game_input_event* event);
   //@}
 
-  static void ParseLibretroName(const std::string& addonName, std::string& emulatorName, std::string& platforms);
+  /*!
+   * \brief Parse the name of a libretro game add-on into its emulator name
+   * and platforms
+   *
+   * \param addonName The name of the add-on, e.g. "Nintendo - SNES / SFC (Snes9x 2002)"
+   *
+   * \return A tuple of two strings:
+   *   - first:  the emulator name, e.g. "Snes9x 2002"
+   *   - second: the platforms, e.g. "Nintendo - SNES / SFC"
+   *
+   * For cores that don't emulate a platform, such as 2048 with the add-on name
+   * "2048", then the emulator name will be the add-on name and platforms
+   * will be empty, e.g.:
+   *   - first:  "2048"
+   *   - second: ""
+   */
+  static std::pair<std::string, std::string> ParseLibretroName(const std::string& addonName);
```

and

```patch
--- a/xbmc/games/addons/GameClient.cpp
+++ b/xbmc/games/addons/GameClient.cpp
@@ -107,7 +107,7 @@ CGameClient::CGameClient(const ADDON::AddonInfoPtr& addonInfo)
   m_bSupportsStandalone =
       addonInfo->Type(AddonType::GAMEDLL)->GetValue(GAME_PROPERTY_SUPPORTS_STANDALONE).asBoolean();
 
-  ParseLibretroName(Name(), m_emulatorName, m_platforms);
+  [m_emulatorName, m_platforms] = ParseLibretroName(Name());
 }
```

But when I compile, I get the error:

```
GameClient.cpp: In constructor ‘KODI::GAME::CGameClient::CGameClient(const AddonInfoPtr&)’:
GameClient.cpp:110:4: error: capture of non-variable ‘KODI::GAME::CGameClient::m_emulatorName’
  110 |   [m_emulatorName, m_platforms] = ParseLibretroName(Name());
      |    ^~~~~~~~~~~~~~
In file included from GameClient.cpp:9:
GameClient.h:244:15: note: ‘std::string KODI::GAME::CGameClient::m_emulatorName’ declared here
  244 |   std::string m_emulatorName;
      |               ^~~~~~~~~~~~~~
```

Does this not work for member variables?

Comment (lrusak) [reply xbmc/games/addons/GameClient.cpp:700]: Try
```
std::tie(m_emulatorName, m_platforms) = ParseLibretroName(Name());
```

**Comment by garbear** [comment]
jenkins build this please

**Pull request merged**